File: 2015/day6/c++/2015.6.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <regex>

using namespace std;

void part1() {
	ifstream input("input.txt");

	string line = "";
	regex pattern(R"((.*) (\d+),(\d+) through (\d+),(\d+))");
	smatch matches;

	string operation = "";
	int x1 = 0;
	int y1 = 0;
	int x2 = 0;
	int y2 = 0;

	vector<vector<bool>> lights(1000, vector<bool>(1000, false));

	while (getline(input, line)) {

		if (!regex_match(line, matches, pattern)) {
			cout << "Invalid command" << endl;
			break;
		}

		operation = matches[1];
		x1 = stoi(matches[2]);
		y1 = stoi(matches[3]);
		x2 = stoi(matches[4]);
		y2 = stoi(matches[5]);

		for (int x = x1; x <= x2; x++) {
			for (int y = y1; y <= y2; y++) {
				if (operation == "turn on") {
					lights[x][y] = true;
				}
				else if (operation == "turn off") {
					lights[x][y] = false;
				}
				else if (operation == "toggle") {
					lights[x][y] = !lights[x][y];
				}
				else {
					cout << "Invalid command" << endl;
					break;
				}
			}
		}
	}

	input.close();

	int lightsOn = 0;

	for (int x = 0; x < lights.size(); x++) {
		for (int y = 0; y < lights[x].size(); y++) {
			if (lights[x][y]) {
				lightsOn++;
			}
		}
	}

	cout << "Part 1 - Lights on: " << lightsOn << endl;
}

void part2() {
	ifstream input("input.txt");

	string line = "";
	regex pattern(R"((.*) (\d+),(\d+) through (\d+),(\d+))");
	smatch matches;

	string operation = "";
	int x1 = 0;
	int y1 = 0;
	int x2 = 0;
	int y2 = 0;

	vector<vector<int>> lights(1000, vector<int>(1000, 0));

	while (getline(input, line)) {

		if (!regex_match(line, matches, pattern)) {
			cout << "Invalid command" << endl;
			break;
		}

		operation = matches[1];
		x1 = stoi(matches[2]);
		y1 = stoi(matches[3]);
		x2 = stoi(matches[4]);
		y2 = stoi(matches[5]);

		for (int x = x1; x <= x2; x++) {
			for (int y = y1; y <= y2; y++) {
				if (operation == "turn on") {
					lights[x][y] += 1;
				}
				else if (operation == "turn off") {
					lights[x][y] = lights[x][y] > 0 ? lights[x][y] - 1 : lights[x][y];
				}
				else if (operation == "toggle") {
					lights[x][y] += 2;
				}
				else {
					cout << "Invalid command" << endl;
					break;
				}
			}
		}
	}

	input.close();

	int totalBrightness = 0;

	for (int x = 0; x < lights.size(); x++) {
		for (int y = 0; y < lights[x].size(); y++) {
			if (lights[x][y]) {
				totalBrightness += lights[x][y];
			}
		}
	}

	cout << "Part 2 - Total Brightness: " << totalBrightness << endl;
}

int main() {
	part1();
	part2();

	return 0;
}

File: 2015/day6/js/2015.6.mjs
import * as readline from "node:readline";
import { createReadStream } from "node:fs";

function part1() {
    const rl = readline.createInterface({
        input: createReadStream("../input.txt"),
        crlfDelay: Infinity // crlfDelay option to recognize all instances of CR LF ('\r\n') in input.txt as a single line break.
    });

    let lights = Array(1000).fill().map(() => Array(1000).fill(false));

    let startX = 0;
    let startY = 0;
    let endX = 0;
    let endY = 0;

    rl.on('line', (line) => {
        if (line === '') return;
        let [, action, x1, y1, x2, y2] = line.match(/(.*) (\d+),(\d+) through (\d+),(\d+)/);

        startX = +x1;
        endX = +x2;
        startY = +y1;
        endY = +y2;

        for (let x = startX; x <= endX; x++) {
            for (let y = startY; y <= endY; y++) {
                switch (action) {
                    case "turn on": lights[x][y] = true; break;
                    case "turn off": lights[x][y] = false; break;
                    case "toggle": lights[x][y] = !lights[x][y]; break;
                }
            }
        }
    });


    rl.on('close', () => {
        let count = 0;
        for (let x = 0; x < lights.length; x++) {
            for (let y = 0; y < lights[x].length; y++) {
                if (lights[x][y] === true) {
                    count++;
                }
            }
        }

        console.log(`Part 1: ${count}`);
    });
}

function part2() {
    const rl = readline.createInterface({
        input: createReadStream("../input.txt"),
        crlfDelay: Infinity // crlfDelay option to recognize all instances of CR LF ('\r\n') in input.txt as a single line break.
    });

    let lights = Array(1000).fill().map(() => Array(1000).fill(0));

    let startX = 0;
    let startY = 0;
    let endX = 0;
    let endY = 0;

    rl.on('line', (line) => {
        if (line === '') return;
        let [, action, x1, y1, x2, y2] = line.match(/(.*) (\d+),(\d+) through (\d+),(\d+)/);

        startX = +x1;
        endX = +x2;
        startY = +y1;
        endY = +y2;

        for (let x = startX; x <= endX; x++) {
            for (let y = startY; y <= endY; y++) {
                switch (action) {
                    case "turn on": lights[x][y] += 1; break;
                    case "turn off": lights[x][y] = lights[x][y] > 0 ? lights[x][y] - 1 : lights[x][y]; break;
                    case "toggle": lights[x][y] += 2; break;
                }
            }
        }
    });


    rl.on('close', () => {
        let totalBrightness = 0;
        for (let x = 0; x < lights.length; x++) {
            for (let y = 0; y < lights[x].length; y++) {
                    totalBrightness+= lights[x][y];
            }
        }

        console.log(`Part 2: ${totalBrightness}`);
    });
}

part1();
part2();